### PR TITLE
#42 Logging now reports the correct number of documents inserted

### PR DESCRIPTION
**Pull request by rjrudin**
Also did some cleanup to reduce duplicated code and extracted some protected methods for more readability. 

Renamed "uriFlowFileMap" to "uuidFlowFileMap", which is the correct name as the keys are FlowFile UUIDs, not MarkLogic URIs.

### CODE REVIEW

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java [hunk 1]
```diff
@@ -226,53 +235,76 @@ public void init(ProcessorInitializationContext context) {
...
         if (serverTransform != null) {
             writeBatcher.withTransform(serverTransform);
         }
+
         Integer threadCount = context.getProperty(THREAD_COUNT).asInteger();
-        if(threadCount != null) {
+        if (threadCount != null) {
             writeBatcher.withThreadCount(threadCount);
         }
-        this.writeBatcher.onBatchSuccess(writeBatch -> {
```
Comment (rjrudin) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: Moved the success/failure code into separate protected methods so it's easier to understand and optionally override.

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java [hunk 2]
```diff
@@ -226,53 +235,76 @@ public void init(ProcessorInitializationContext context) {
...
-                }
-            }
-        }).onBatchFailure((writeBatch, throwable) -> {
+
+        dataMovementManager.startJob(writeBatcher);
+    }
+
+    protected void onBatchSuccess(WriteBatch writeBatch) {
+        if (writeBatch.getItems().length > 0) {
+            getLogger().info("Inserted {} documents from batch number {}",
```
Comment (rjrudin) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: This is the actual logging fix - logging it after success and including the correct number of documents and the batch number too (which I find helpful).

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java [hunk 3]
```diff
@@ -226,53 +235,76 @@ public void init(ProcessorInitializationContext context) {
...
         }
+
         Integer threadCount = context.getProperty(THREAD_COUNT).asInteger();
-        if(threadCount != null) {
+        if (threadCount != null) {
             writeBatcher.withThreadCount(threadCount);
         }
-        this.writeBatcher.onBatchSuccess(writeBatch -> {
-            if (writeBatch.getItems().length > 0) {
-                ProcessSession session = getFlowFileInfoForWriteEvent(writeBatch.getItems()[0]).session;
```
Comment (rjrudin) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: Moved this into a new protected method - transferToBatchSuccessRelationship - for readability.

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogicRecord.java [hunk 4]
```diff
@@ -207,35 +208,11 @@ protected WriteEvent buildWriteEvent(
             final FlowFile flowFile,
             String uri,
             final BytesHandle contentHandle,
-            final Map<String, String> additionalAttributes
+            final DocumentMetadataHandle metadata
     ) {
-        final String prefix = context.getProperty(URI_PREFIX).evaluateAttributeExpressions(flowFile).getValue();
```
Comment (rjrudin) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogicRecord.java]: This was mostly duplicated code - moved it to protected methods in PutMarkLogic for reuse.

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogicRecord.java [hunk 5]
```diff
@@ -154,15 +157,14 @@ public final void onTrigger(ProcessContext context, ProcessSession session) thro
...
-                    final WriteResult writeResult = writer.write(record);
-                    additionalAttributes = writeResult.getAttributes();
+                    writer.write(record);
                     writer.flush();
                     BytesHandle bytesHandle = new BytesHandle().with(baos.toByteArray());
                     final String uriKey = uriFieldName == null ? UUID.randomUUID().toString() : record.getAsString(uriFieldName);
-                    WriteEvent writeEvent = buildWriteEvent(context, session, flowFile, uriKey, bytesHandle, additionalAttributes);
+                    WriteEvent writeEvent = buildWriteEvent(context, session, flowFile, uriKey, bytesHandle, metadata);
                     this.addWriteEvent(writeBatcher, writeEvent);
+                    added++;
```
Comment (rjrudin) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogicRecord.java]: This fixes the issue with zero being reported to the provenance report every time.

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java [hunk 6]
```diff
@@ -339,6 +372,24 @@ protected void addWriteEvent(WriteBatcher writeBatcher, WriteEvent writeEvent) {
...
+        DocumentMetadataHandle metadata = buildMetadataHandle(context, flowFile, context.getProperty(COLLECTIONS), context.getProperty(PERMISSIONS));
+        final byte[] content = new byte[(int) flowFile.getSize()];
+        session.read(flowFile, inputStream -> StreamUtils.fillBuffer(inputStream, content));
+
+        BytesHandle handle = new BytesHandle(content);
+        applyPropertiesToContentHandle(context, uri, handle);
+
+        registerFlowFileInMap(flowFile, session);
+
+        return new WriteEventImpl()
```
Comment (ehennum) [inline nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: I'm a little concerned that the NiFi implementation is reaching into the DMSDK implementation, but let's not pursue that now.

File: nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java (hunk 1 above)
Comment (ehennum) [reply nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: Much easier to understand, thanks.

Comment (rjrudin) [reply nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: @ehennum I'm planning on adding at least unit tests this week for all of these PRs. Will see about integration tests.

Comment (ehennum) [reply nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/PutMarkLogic.java]: Thanks, @rjrudin . If the pull request can provide an appropriate level of coverage (no need for tests for test's sake), that's greatly appreciated.

**Comment by rjrudin** [comment]
Closing this, will submit a PR with only the fix in it, no cleanup.

**Pull request closed without merging**